ISSUE-107: fix out of memory when fetching Top objects with lots of children

DiegoPino opened the pull request:
# What?

See #107. 
- Adds adaptive Range/offset based on what was the last number of Top objects v/s Children Rations
- Copies CMODEL mapping between Children/Top objects during setup
- Validates UUID Mapping v/s Node to Node Mapping (to avoid those being mixed up)
- Multiple cleanups, fixes Automatic UUID for Plugins that run via Batch (e.g Solr)

@karomabiles and @alliomeria even if this is safer, I feel it just became a bit slower so will need you both to check with me during our call the same collection being harvested via your current systems v/s with this patch. I can do a pre-test just to be sure of course and make you loose less time

aksm commented on src/Form/AmiMultiStepIngest.php:
```diff
@@ -150,12 +150,12 @@ public function buildForm(array $form, FormStateInterface $form_state) {
-        '#options' => $webform,
-        '#description' => $this->t('Columns are casted to ADO metadata (JSON) by passing/validating Data through an existing Webform'),
-      ];*/
+      /* $element_conditional['webform'] =[
+         '#type' => 'select',
+         '#title' => $this->t('Webform'),
+         '#options' => $webform,
+         '#description' => $this->t('Columns are casted to ADO metadata (JSON) by passing/validating Data through an existing Webform'),
+       ];*/
```
```suggestion
      /**
       * $element_conditional['webform'] = [
       *   '#type' => 'select',
       *   '#title' => $this->t('Webform'),
       *   '#options' => $webform,
       *   '#description' => $this->t('Columns are casted to ADO metadata (JSON) by passing/validating Data through an existing Webform'),
       * ];
       */
```

aksm commented on src/Form/AmiMultiStepIngest.php:
```diff
@@ -533,12 +576,18 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
 
-        // We should probably add the UUIDs here right now.
-        $uuid_key = isset($amisetdata->adomapping['uuid']['uuid']) && !empty($amisetdata->adomapping['uuid']['uuid']) ? $amisetdata->adomapping['uuid']['uuid'] : 'node_uuid';
+        /* remove from the mappings any column used as file source */
```
```suggestion
        /** remove from the mappings any column used as file source */
```

DiegoPino replied:
I will leave this one as it is. Single commends can have /* */ or even //

aksm commented on src/Form/AmiMultiStepIngest.php:
```diff
@@ -533,12 +576,18 @@ public function submitForm(array &$form, FormStateInterface $form_state) {
-          $fileid = $this->AmiUtilityService->csv_save($data, $uuid_key);
+          // We now pass also if auto UUID was chosen or not.
+          $fileid = $this->AmiUtilityService->csv_save($data, $uuid_key, $amisetdata->adomapping['autouuid'] ?? FALSE);
         } else {
```
```suggestion
        }
        else {
```

aksm commented on src/Plugin/ImporterAdapter/SolrImporter.php:
```diff
@@ -122,14 +122,23 @@ public function interactiveForm(array $parents, FormStateInterface $form_state):
       $form_state->setValue(['pluginconfig','ready'], FALSE);
     }
 
+    /* Calculate Max timeout based on PHP Settings */
```
```suggestion
    /** Calculate Max timeout based on PHP Settings */
```

aksm commented on src/Plugin/ImporterAdapter/SolrImporter.php:
```diff
@@ -231,12 +245,14 @@ public function interactiveForm(array $parents, FormStateInterface $form_state):
       ],
     ];
-
+    /* @TODO let's tell the user why it is not ready? */
+    /* Also todo, move mappings already done between children/parents/children */
+    /* That way previous decisions on any of each sides will stick on the other */
```
```suggestion
    /** 
     * @TODO let's tell the user why it is not ready?
     * Also todo, move mappings already done between children/parents/children
     * That way previous decisions on any of each sides will stick on the other
     */
```

aksm commented on src/Plugin/ImporterAdapter/SolrImporter.php:
```diff
@@ -445,7 +470,15 @@ public function getInfo(array $config, FormStateInterface $form_state, $page = 0
+        $escaped = '('. $escaped .')';
+
         $query->setQuery('*:*');
         $query->createFilterQuery('collection_members')->setQuery('RELS_EXT_isMemberOfCollection_uri_s:'.$escaped .' OR RELS_EXT_isMemberOf_uri_s:'.$escaped );
```
```suggestion
        $query->createFilterQuery('collection_members')->setQuery('RELS_EXT_isMemberOfCollection_uri_s:' . $escaped . ' OR RELS_EXT_isMemberOf_uri_s:' . $escaped);
```

aksm commented on src/Plugin/ImporterAdapter/SolrImporter.php:
```diff
@@ -607,7 +640,7 @@ public function getInfo(array $config, FormStateInterface $form_state, $page = 0
-        && count($cmodel_children) >= 1) || $form_state->getValue(['pluginconfig', 'solarium_mapping', 'collapse']) == TRUE) {
+          && count($cmodel_children) >= 1) || $form_state->getValue(['pluginconfig', 'solarium_mapping', 'collapse']) == TRUE) {
         $form_state->setValue(['pluginconfig', 'ready'], TRUE);
       } else {
```
```suggestion
      }
      else {
```

aksm commented on src/Plugin/ImporterAdapter/SolrImporter.php:
```diff
@@ -682,22 +740,39 @@ public function getData(array $config, $page = 0, $per_page = 20): array {
+        $escaped = '('. $escaped .')';
         $query->setQuery('*:*');
+
         $query->createFilterQuery('collection_members')->setQuery('RELS_EXT_isMemberOfCollection_uri_s:'.$escaped .' OR RELS_EXT_isMemberOf_uri_s:'.$escaped );
```
```suggestion
        $query->createFilterQuery('collection_members')->setQuery('RELS_EXT_isMemberOfCollection_uri_s:' . $escaped . ' OR RELS_EXT_isMemberOf_uri_s:' . $escaped);
```

aksm commented on src/Plugin/ImporterAdapter/SolrImporter.php:
```diff
@@ -778,16 +853,18 @@ public function getData(array $config, $page = 0, $per_page = 20): array {
+          // If deep traverse is enabled we should also mark collections here?
+          // Call getData recursevely?
+
         } catch (\Exception $exception) {
```
```suggestion
        }
        catch (\Exception $exception) {
```

aksm commented on src/Plugin/ImporterAdapter/SolrImporter.php:
```diff
@@ -903,8 +994,6 @@ protected function getDataChildren(array $config, SolariumClient $client, string
     $resultset = $client->select($query);
-    // display the total number of documents found by solr
-    error_log('NumFound: ' . $resultset->getNumFound());
     if ($resultset->getNumFound() == 0) { return []; }
```
```suggestion
    if ($resultset->getNumFound() == 0) {
      return [];
    }
```

aksm commented on src/Plugin/ImporterAdapter/SolrImporter.php:
```diff
@@ -1106,17 +1238,31 @@ public static function fetchBatch(array $config, ImporterPluginAdapterInterface
+      $data = $plugin_instance->getData($config, $nextoffset,
         $next_increment);
+
+      if (isset($data['nextforcedoffset']) && $data['nextforcedoffset'] !== NULL ) {
```
```suggestion
      if (isset($data['nextforcedoffset']) && $data['nextforcedoffset'] !== NULL) {
```

DiegoPino commented:
@aksm thanks for your review

Any reason why we need to use

` /**comments  */`

instead of 
`/* comments */`

Just checking my Drupal Coding standards knowledge. Was not aware that is a requirement? Please let me know

aksm commented:
> @aksm thanks for your review
> 
> Any reason why we need to use
> 
> ` /**comments */`
> 
> instead of `/* comments */`
> 
> Just checking my Drupal Coding standards knowledge. Was not aware that is a requirement? Please let me know

@DiegoPino Sorry, I misunderstood the docs. I thought all of them had to start that way, but I see now that it's just docblocks.

DiegoPino commented:
@aksm no worries. DCS are a lot to digest. Please check if I addressed all your concerns. If so we can move forward and approve and keep walking into other issues. Before merging I need to compare performance with @karomabiles and @alliomeria. I felt a small harvest last night went a bit slow. I can't see why/where it could have been slowed down so it might be just my internet, but still better to compare. Thanks a lot for reviewing!

aksm commented:
@DiegoPino Looks good to me. Was only scanning for (my limited understanding of) DCS and anything obvious that might jump out at me. Thanks!

DiegoPino commented:
@aksm if it looks good you need to approve it!

DiegoPino commented:
There might be some Drupal 10.0 deprecations in some parts of AMI. Will deal with those in another pull. In the meantime merging this and we go from there. Thanks
